api/clients/v2/disperser_client.go

```diff
@@ -390,7 +393,137 @@ func (c *disperserClient) GetPaymentState(ctx context.Context) (*disperser_rpc.G
+		return nil, fmt.Errorf("legacy payment state received from disperser does not contain global params")
+	}
+	// Convert PaymentGlobalParams to PaymentVaultParams
+	var paymentVaultParams *disperser_rpc.PaymentVaultParams
+	{
+		paymentVaultParams = &disperser_rpc.PaymentVaultParams{
+			QuorumPaymentConfigs:  make(map[uint32]*disperser_rpc.PaymentQuorumConfig),
+			QuorumProtocolConfigs: make(map[uint32]*disperser_rpc.PaymentQuorumProtocolConfig),
+			OnDemandQuorumNumbers: legacyReply.PaymentGlobalParams.OnDemandQuorumNumbers,
+		}
+
+		// Apply the global params to all quorums mentioned in on-demand quorum numbers
+		// If no on-demand quorums specified, apply to quorum 0 as default
+		quorums := legacyReply.PaymentGlobalParams.OnDemandQuorumNumbers
+		if len(quorums) == 0 {
+			return nil, fmt.Errorf("no on-demand quorums specified in legacy payment state received from disperser")
+		}
+
+		for _, quorumID := range quorums {
+			paymentVaultParams.QuorumPaymentConfigs[quorumID] = &disperser_rpc.PaymentQuorumConfig{
+				ReservationSymbolsPerSecond: 0, // Not available in legacy format
+				OnDemandSymbolsPerSecond:    legacyReply.PaymentGlobalParams.GlobalSymbolsPerSecond,
+				OnDemandPricePerSymbol:      legacyReply.PaymentGlobalParams.PricePerSymbol,
+			}
+
+			paymentVaultParams.QuorumProtocolConfigs[quorumID] = &disperser_rpc.PaymentQuorumProtocolConfig{
+				MinNumSymbols: legacyReply.PaymentGlobalParams.MinNumSymbols,
+				// ReservationAdvanceWindow is not used offchain at the moment so it's okay to set to any value.
+				// It was added for consistency with the onchain data structure but get removed in the future.
+				ReservationAdvanceWindow:   legacyReply.PaymentGlobalParams.ReservationWindow,
+				ReservationRateLimitWindow: legacyReply.PaymentGlobalParams.ReservationWindow,
+				OnDemandRateLimitWindow:    0, // Not available in legacy format
+				OnDemandEnabled:            len(legacyReply.PaymentGlobalParams.OnDemandQuorumNumbers) > 0,
+			}
+		}
+	}
+
+	// Convert aggregated period records to per-quorum format
+	periodRecords := make(map[uint32]*disperser_rpc.PeriodRecords)
+	// If no period records are available, return empty map
+	// This is possible if no reservations exist for the account
+	if len(legacyReply.PeriodRecords) > 0 {
+		// Apply the same period records to all relevant quorums
+		quorums := []uint32{0} // Default to quorum 0
+		if legacyReply.Reservation != nil && len(legacyReply.Reservation.QuorumNumbers) > 0 {
+			quorums = legacyReply.Reservation.QuorumNumbers
+		}
+
+		for _, quorumID := range quorums {
+			periodRecords[quorumID] = &disperser_rpc.PeriodRecords{
+				Records: legacyReply.PeriodRecords,
+			}
+		}
+	}
+
+	// Convert aggregated reservation to per-quorum format
+	reservations := make(map[uint32]*disperser_rpc.QuorumReservation)
+	// If no reservation is available, return empty map
+	// This is possible if no reservations exist for the account
+	if legacyReply.Reservation != nil {
```

Comment:
I think the logic of this method could be simplified with the following strategy:

1. Before handling anything to do with a reservation, check `legacyReply.Reservation == nil`. If it is `nil`, you can immediately return your `GetPaymentStateForAllQuorumsReply`, with some empty fields.
2. Assert that all expected fields within the reservation object are non-nil.
3. Then, you are free to do all your reservation logic without having to wonder about whether there are actually any reservations, and if any field might be unexpectedly nil